spr/src/commands/diff.rs

```diff
@@ -41,12 +42,54 @@ pub struct DiffOptions {
     #[clap(long, short = 'm')]
     message: Option<String>,
 
+    /// Which commits in the branch should be created/updated. This can be a
+    /// revspec such as HEAD~4..HEAD~1 or just one commit like HEAD~7.
+    #[clap(long, short = 'r')]
+    refs: Option<String>,
+
     /// Submit this commit as if it was cherry-picked on master. Do not base it
     /// on any intermediate changes between the master branch and this commit.
     #[clap(long)]
     cherry_pick: bool,
 }
 
+fn get_oids(refs: &String, repo: &git2::Repository) -> Result<HashSet<Oid>> {
+    // refs might be a single (eg 012345abc or HEAD) or a range (HEAD~4..HEAD~2)
+    let revspec = repo.revparse(&refs)?;
+    if revspec.mode().contains(git2::RevparseMode::MERGE_BASE) {
+        // this is ok, as other code doesn't handle merges
+        return Err(Error::new("Unable to cope with merge_base --refs"));
+    } else if revspec.mode().contains(git2::RevparseMode::SINGLE) {
+        // simple case, just return the id
+        return Ok(HashSet::from([revspec.from().unwrap().id()]));
+    } else if revspec.mode().contains(git2::RevparseMode::RANGE) {
+        // If it's a range we have to walk from `from` to `to` grabbing oids
+        let mut to = revspec.to().ok_or(Error::new("Unexpectedly no to id in range"))?.id();
+        let from = revspec.from().ok_or(Error::new("Unexpectedly no from id in range"))?.id();
+        let mut ret = HashSet::new();
+        loop {
+            ret.insert(to);
+            if to == from {
+                // finished the walk
+                break;
+            }
+            let commit_to = repo.find_commit(to)?;
+            let mut oid_iter = commit_to.parent_ids();
+            let next_oid = oid_iter.next();
+            if let Some(_) = oid_iter.next() {
+                return Err(Error::new("Unexpectedly had multiple parents for commit"));
+            }
+            match next_oid {
+                None => return Err(Error::new("Unexpectedly no parent id for commit")),
+                Some(id) => to = id,
+            }
+        }
+        return Ok(ret);
+    } else {
+        return Err(Error::new("Unable to cope with this type of revspec for --refs"));
+    }
```

Comment:
I don't think you have to implement all this yourself. `git2` has a [`Revwalk`](https://docs.rs/git2/latest/git2/struct.Revwalk.html) class to do this kind of thing.

Something along the lines of
```
let to = revspec.to().ok_or(Error::new("Unexpectedly no to id in range"))?.id();
let from = revspec.from().ok_or(Error::new("Unexpectedly no from id in range"))?.id();

let mut walk = repo.revwalk()?;
walk.push(to)?;
walk.hide(from)?;
walk.map(|r| Ok(r?)).collect()
```

The iterator stuff in Rust is super nice. The iterator that `walk.map` returns yields `Result<Oid, git2::Error>` items. The `collect` method of the iterator trait is able to collect `Result<T, E>` into `Result<HashSet<T>, E>`. But since the error type that git2 gives us is different from the one we return here, we need to put in that `map(|r| Ok(r?))` thing, where the `?` operator does the conversion. (Maybe there is a more idiomatic way to the error mapping.)

The slight difference to your implementation is that the resulting set won't include them `from` commit, but I think that's more aligned with Git anyway. (E.g. `git log HEAD~3..HEAD` will show 3 commits and won't include `HEAD~3`)

Reply:
yeah - thats way neater